drivers/display/display_framebuf.c

```diff
@@ -0,0 +1,251 @@
+ * simplicity. There's a lot of room for performance improvement.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <display.h>
+#include <display/framebuf.h>
+#include <string.h>
+
+#define GREY 0x00C0C0C0
+#define BLACK 0x00000000
+
+static int framebuf_blanking_on(const struct device *dev)
+{
+	return -ENOTSUP;
+}
+
+static int framebuf_blanking_off(const struct device *dev)
+{
+	return -ENOTSUP;
+}
+
+static void *framebuf_get_framebuffer(const struct device *dev)
+{
+	return NULL;
+}
+
+static int framebuf_set_brightness(const struct device *dev,
+				   const u8_t brightness)
+{
+	return -ENOTSUP;
+}
+
+static int framebuf_set_contrast(const struct device *dev,
+				 const u8_t contrast)
+{
+	return -ENOTSUP;
+}
+
+static int framebuf_set_pixel_format(const struct device *dev,
+				     const enum display_pixel_format format)
+{
+	struct framebuf_dev_data *data = FRAMEBUF_DATA(dev);
+
+	switch (format) {
+	case PIXEL_FORMAT_ARGB_8888:
+	case PIXEL_FORMAT_MONO01:
+	case PIXEL_FORMAT_MONO10:
+		data->format = format;
+		return 0;
+	default:
+		return -ENOTSUP;
+	}
+}
+
+static int framebuf_set_orientation(const struct device *dev,
+				    const enum display_orientation orientation)
+{
+	switch (orientation) {
+	case DISPLAY_ORIENTATION_NORMAL:
+		return 0;
+	default:
+		return -ENOTSUP;
+	}
+}
+
+static void framebuf_get_capabilities(const struct device *dev,
+				      struct display_capabilities *caps)
+{
+	struct framebuf_dev_data *data;
+
+	data = FRAMEBUF_DATA(dev);
+	caps->x_resolution = data->width;
+	caps->y_resolution = data->height;
+	caps->supported_pixel_formats = PIXEL_FORMAT_ARGB_8888 |
+					PIXEL_FORMAT_MONO01 |
+					PIXEL_FORMAT_MONO10;
+	caps->screen_info = 0;
+	caps->current_pixel_format = data->format;
+	caps->current_orientation = DISPLAY_ORIENTATION_NORMAL;
+}
+
+/*
+ * simple bit manipulation functions for display buffer read/write.
+ *
+ * we can't use the system-defined ones because they work on words,
+ * not bytes, which would introduce a dependency on endianness.
+ */
+
+static inline int test_bit(const void *buf, int bit)
+{
+	const u8_t *buf8 = buf;
+
+	return buf8[bit >> 3] & (1 << (7 - (bit & 7)));
+}
+
+static inline void set_bit(void *buf, int bit)
+{
+	u8_t *buf8 = buf;
+
+	buf8[bit >> 3] |= (1 << (7 - (bit & 7)));
+}
+
+static inline void reset_bit(void *buf, int bit)
+{
+	u8_t *buf8 = buf;
+
+	buf8[bit >> 3] &= ~(1 << (7 - (bit & 7)));
+}
+
+static int framebuf_write(const struct device *dev, const u16_t x,
+			  const u16_t y,
+			  const struct display_buffer_descriptor *desc,
+			  const void *buf)
+{
+	struct framebuf_dev_data *data = FRAMEBUF_DATA(dev);
+	u32_t *dst = data->buffer;
+	const u32_t *src;
+	int row, col, bit;
+	u32_t set_color, reset_color;
+
+	dst += x;
+	dst += (y * data->pitch);
+
+	switch (data->format) {
+	case PIXEL_FORMAT_ARGB_8888:
+		src = buf;
+		for (row = 0; row < desc->height; ++row) {
+			memcpy(dst, src, desc->width * sizeof(u32_t));
```

Comment:
memcpy on top of the single pixel word is likewise sort of a performance disaster in graphics code. :)

Reply:
Still lost :(